c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -191,6 +310,24 @@ impl<'tcx> TypeChecker<'tcx, '_> {
             let point = self.current_point(SubPoint::Mid);
             self.facts.subset_base.push((rv_origin, pl_origin, point));
         }
+
+        for (pl_lty, rv_lty) in pl_lty.iter().zip(rv_lty.iter()) {
+            let pl_origin_params = pl_lty.label.origin_params;
+            let rv_origin_params = rv_lty.label.origin_params;
+            if let (Some(pl_origin_params), Some(rv_origin_params)) =
+                (pl_origin_params, rv_origin_params)
+            {
+                assert_eq!(pl_origin_params.len(), rv_origin_params.len());
+                for (pl_origin_param, rv_origin_param) in
+                    pl_origin_params.iter().zip(rv_origin_params.iter())
+                {
+                    let point = self.current_point(SubPoint::Mid);
+                    self.facts
+                        .subset_base
+                        .push((rv_origin_param.1, pl_origin_param.1, point));
+                }
+            }
+        }
```

Comment:
```suggestion
        let mut add_mid_point = |pl: Origin, rv: Origin| {
            let point = self.current_point(SubPoint::Mid);
            self.facts.subset_base.push((rv, pl, point));
        };

        if let (Some(pl), Some(rv)) = (pl_lty.label.origin, rv_lty.label.origin) {
            add_mid_point(pl, rv);
        }

        for (pl_lty, rv_lty) in zip(pl_lty.iter(), rv_lty.iter()) {
            if let (Some(pl), Some(rv)) = (pl_lty.label.origin_params, rv_lty.label.origin_params) {
                assert_eq!(pl.len(), rv.len());
                for (&(_, pl_origin), &(_, rv_origin)) in zip(pl, rv) {
                    add_mid_point(rv_origin, pl_origin);
                }
            }
        }
```

This should also replace the previous part of the function, too, but GitHub won't let me highlight those lines in the review.  This seems easier to read/understand to me, with
- the refactored out `add_mid_point`
- the free `zip` function rather than the method
- more concise variable names since they're repeated so often in close proximity here

Reply:
Looks good, with two nitpicks:
1. `add_mid_point` should be called something like `add_subset_base` - the fact that it uses `SubPoint::Mid` is not the most important thing about it
2. Currently, the first two variables named `pl` (and `rv`) have the same type (`Origin`), but the third has a different type (`&[(OriginParam, Origin)]`).  I think the second loop should name the three levels of variables something like `pl_lty`, `pl_origins`, `pl` so that the use of `pl` remains consistent within this function.

Reply:
addressed in 356d7b18